File: src/aircraft_movement.cpp
#include <winsock2.h>
#include <windows.h>
#include "SimConnect.h"
#include <iostream>
#include <thread>
#include <atomic>
#include <mutex>
#include "json.hpp"
#include "httplib.h"

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

// External declarations from simconnect_bridge.cpp
extern HANDLE hSimConnect;
extern std::atomic<bool> quit;

// Data definition ID for aircraft movement
#define DEFINITION_3 3
// Add new definition for cockpit view reset
#define DEFINITION_4 4

// Aircraft movement control structure (6 metrics including cockpit camera zoom and pitch)
struct AircraftMovementData {
    double latitude;
    double longitude;
    double altitude;
    double heading;
    double pitch;
    double cockpit_camera_zoom;
};
// Structure for cockpit view reset
struct CockpitViewResetData {
    int32_t reset;
};

// Global variables for aircraft movement
std::atomic<bool> movement_thread_quit{false};
std::mutex movement_mutex;
AircraftMovementData current_movement_data = {0.0, 0.0, 0.0, 0.0, 0.0, 1.0};
bool has_movement_data = false;

// Function to set up the aircraft movement data definition
void SetupAircraftMovementDefinition() {
    if (hSimConnect == nullptr) {
        std::cerr << "SimConnect handle is null, cannot set up aircraft movement definition" << std::endl;
        return;
    }

    // Define aircraft movement data structure (5 metrics including cockpit camera zoom)
    SimConnect_AddToDataDefinition(hSimConnect, DEFINITION_3, "PLANE LATITUDE", "degrees", SIMCONNECT_DATATYPE_FLOAT64);
    SimConnect_AddToDataDefinition(hSimConnect, DEFINITION_3, "PLANE LONGITUDE", "degrees", SIMCONNECT_DATATYPE_FLOAT64);
    SimConnect_AddToDataDefinition(hSimConnect, DEFINITION_3, "PLANE ALTITUDE", "ft", SIMCONNECT_DATATYPE_FLOAT64);
    SimConnect_AddToDataDefinition(hSimConnect, DEFINITION_3, "PLANE HEADING DEGREES TRUE", "radian", SIMCONNECT_DATATYPE_FLOAT64);
    SimConnect_AddToDataDefinition(hSimConnect, DEFINITION_3, "PLANE PITCH DEGREES", "radian", SIMCONNECT_DATATYPE_FLOAT64);
    SimConnect_AddToDataDefinition(hSimConnect, DEFINITION_3, "COCKPIT CAMERA ZOOM", "percentage", SIMCONNECT_DATATYPE_FLOAT64);

    // Define cockpit view reset (DEFINITION_4)
    SimConnect_AddToDataDefinition(hSimConnect, DEFINITION_4, "CAMERA REQUEST ACTION", "number", SIMCONNECT_DATATYPE_INT32);
    
    std::cout << "Aircraft movement data definition (DEFINITION_3) set up successfully" << std::endl;
}

// Function to set aircraft position using SimConnect
void SetAircraftPosition(const AircraftMovementData& data) {
    if (hSimConnect == nullptr) {
        std::cerr << "SimConnect handle is null, cannot set aircraft position" << std::endl;
        return;
    }

    // Convert heading and pitch from degrees to radians for SimConnect
    double heading_rad = data.heading * M_PI / 180.0;
    double pitch_rad = data.pitch * M_PI / 180.0;

    // Create the data structure for DEFINITION_3
    AircraftMovementData simconnect_data = {
        data.latitude,      // latitude in degrees
        data.longitude,     // longitude in degrees
        data.altitude,      // altitude in feet
        heading_rad,        // heading in radians
        pitch_rad,          // pitch in radians
        data.cockpit_camera_zoom  // cockpit camera zoom ratio
    };

    // Set the data on the user aircraft (SimObject ID 1)
    HRESULT hr = SimConnect_SetDataOnSimObject(hSimConnect, DEFINITION_3, SIMCONNECT_OBJECT_ID_USER, 0, 0, sizeof(AircraftMovementData), &simconnect_data);
    
    if (SUCCEEDED(hr)) {
        // std::cout << "Aircraft position set: lat=" << data.latitude << ", lon=" << data.longitude << ", alt=" << data.altitude << "ft, hdg=" << data.heading << ", zoom=" << data.cockpit_camera_zoom << std::endl;
    } else {
        std::cerr << "Failed to set aircraft position. Error: 0x" << std::hex << hr << std::endl;
    }
}

// Function to send cockpit view reset
void SendCockpitViewReset(int32_t reset_value) {
    if (hSimConnect == nullptr) {
        std::cerr << "SimConnect handle is null, cannot send cockpit view reset" << std::endl;
        return;
    }
    CockpitViewResetData reset_data = { reset_value };
    HRESULT hr = SimConnect_SetDataOnSimObject(hSimConnect, DEFINITION_4, SIMCONNECT_OBJECT_ID_USER, 0, 0, sizeof(CockpitViewResetData), &reset_data);
    if (SUCCEEDED(hr)) {
        std::cout << "Cockpit view reset sent: reset=" << reset_value << std::endl;
    } else {
        std::cerr << "Failed to send cockpit view reset. Error: 0x" << std::hex << hr << std::endl;
    }
}

// Function to process movement data from the queue
void ProcessMovementData() {
    while (!movement_thread_quit) {
        AircraftMovementData data_to_process;
        bool has_data = false;

        // Check if we have new movement data
        {
            std::lock_guard<std::mutex> lock(movement_mutex);
            if (has_movement_data) {
                data_to_process = current_movement_data;
                has_data = true;
                has_movement_data = false; // Mark as processed
            }
        }

        // Process the movement data if available
        if (has_data) {
            SetAircraftPosition(data_to_process);
        }

        // Sleep for a short interval before checking again
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }
}

// HTTP server thread for aircraft movement control
void AircraftMovementServerThread() {
    httplib::Server svr;

    // Set default CORS headers
/*   svr.set_default_headers({
        {"Access-Control-Allow-Origin", "*"},
        {"Access-Control-Allow-Methods", "PUT, GET, POST, OPTIONS"},
        {"Access-Control-Allow-Headers", "*"},
        {"Access-Control-Max-Age", "86400"}
    });*/

    // CORS preflight handler for movement endpoint
    svr.Options("/move", [](const httplib::Request&, httplib::Response& res) {
        res.set_header("Access-Control-Allow-Origin", "*");
        res.set_header("Access-Control-Allow-Methods", "PUT, GET, POST, OPTIONS");
        res.set_header("Access-Control-Allow-Headers", "*");
        res.set_header("Access-Control-Max-Age", "86400");
        res.status = 204; // No Content
    });

    // POST endpoint to receive aircraft movement commands
    svr.Post("/move", [](const httplib::Request& req, httplib::Response& res) {
        // Set CORS headers first, before any processing
        res.set_header("Access-Control-Allow-Origin", "*");
        res.set_header("Access-Control-Allow-Methods", "PUT, GET, POST, OPTIONS");
        res.set_header("Access-Control-Allow-Headers", "*");
        res.set_header("Access-Control-Max-Age", "86400");
        
        try {
            // Parse JSON from request body
            nlohmann::json json_data = nlohmann::json::parse(req.body);
            
            // Validate required fields
            if (!json_data.contains("lat") || !json_data.contains("lon") || 
                !json_data.contains("alt") || !json_data.contains("heading") || 
                !json_data.contains("pitch") ||
                !json_data.contains("cockpit_camera_zoom")) {
                res.status = 400;
                res.set_content("{\"error\": \"Missing required fields: lat, lon, alt, heading, pitch, cockpit_camera_zoom\"}", "application/json");
                return;
            }

            // Extract and validate data
            double lat = json_data["lat"].get<double>();
            double lon = json_data["lon"].get<double>();
            double alt = json_data["alt"].get<double>();
            double heading = json_data["heading"].get<double>();
            double pitch = json_data["pitch"].get<double>();
            double cockpit_camera_zoom = json_data["cockpit_camera_zoom"].get<double>();

            // Basic validation
            if (lat < -90.0 || lat > 90.0) {
                res.status = 400;
                res.set_content("{\"error\": \"Latitude must be between -90 and 90 degrees\"}", "application/json");
                return;
            }

            if (lon < -180.0 || lon > 180.0) {
                res.status = 400;
                res.set_content("{\"error\": \"Longitude must be between -180 and 180 degrees\"}", "application/json");
                return;
            }

            if (alt < -1000.0 || alt > 100000.0) {
                res.status = 400;
                res.set_content("{\"error\": \"Altitude must be between -1000 and 100000 feet\"}", "application/json");
                return;
            }

            if (heading < 0.0 || heading > 360.0) {
                res.status = 400;
                res.set_content("{\"error\": \"Heading must be between 0 and 360 degrees\"}", "application/json");
                return;
            }

            if (pitch < -90.0 || pitch > 90.0) {
                res.status = 400;
                res.set_content("{\"error\": \"Pitch must be between -90 and 90 degrees\"}", "application/json");
                return;
            }

            if (cockpit_camera_zoom < 0.0 || cockpit_camera_zoom > 100.0) {
                res.status = 400;
                res.set_content("{\"error\": \"Cockpit camera zoom must be between 0.0 and 100.0\"}", "application/json");
                return;
            }

            // Store the movement data for processing
            {
                std::lock_guard<std::mutex> lock(movement_mutex);
                current_movement_data.latitude = lat;
                current_movement_data.longitude = lon;
                current_movement_data.altitude = alt;
                current_movement_data.heading = heading;
                current_movement_data.pitch = pitch;
                current_movement_data.cockpit_camera_zoom = cockpit_camera_zoom;
                has_movement_data = true;
            }

            // Return success response
            nlohmann::json response = {
                {"status", "success"},
                {"message", "Aircraft movement command received"},
                {"data", {
                    {"lat", lat},
                    {"lon", lon},
                    {"alt", alt},
                    {"heading", heading},
                    {"pitch", pitch},
                    {"cockpit_camera_zoom", cockpit_camera_zoom}
                }}
            };

            res.set_content(response.dump(), "application/json");

        } catch (const std::exception& e) {
            res.status = 400;
            nlohmann::json error_response = {
                {"error", "Invalid JSON format"},
                {"details", e.what()}
            };
            res.set_header("Access-Control-Allow-Origin", "*");
            res.set_header("Access-Control-Allow-Methods", "PUT, GET, POST, OPTIONS");
            res.set_header("Access-Control-Allow-Headers", "*");
            res.set_content(error_response.dump(), "application/json");
        }
    });

    // PUT endpoint to receive aircraft movement commands (same as POST)
    svr.Put("/move", [](const httplib::Request& req, httplib::Response& res) {
        // Set CORS headers first, before any processing
        res.set_header("Access-Control-Allow-Origin", "*");
        res.set_header("Access-Control-Allow-Methods", "PUT, GET, POST, OPTIONS");
        res.set_header("Access-Control-Allow-Headers", "*");
        res.set_header("Access-Control-Max-Age", "86400");
        
        try {
            // Parse JSON from request body
            nlohmann::json json_data = nlohmann::json::parse(req.body);
            
            // Validate required fields
            if (!json_data.contains("lat") || !json_data.contains("lon") || 
                !json_data.contains("alt") || !json_data.contains("heading") || 
                !json_data.contains("pitch") ||
                !json_data.contains("cockpit_camera_zoom")) {
                res.status = 400;
                res.set_content("{\"error\": \"Missing required fields: lat, lon, alt, heading, pitch, cockpit_camera_zoom\"}", "application/json");
                return;
            }

            // Extract and validate data
            double lat = json_data["lat"].get<double>();
            double lon = json_data["lon"].get<double>();
            double alt = json_data["alt"].get<double>();
            double heading = json_data["heading"].get<double>();
            double pitch = json_data["pitch"].get<double>();
            double cockpit_camera_zoom = json_data["cockpit_camera_zoom"].get<double>();

            // Basic validation
            if (lat < -90.0 || lat > 90.0) {
                res.status = 400;
                res.set_content("{\"error\": \"Latitude must be between -90 and 90 degrees\"}", "application/json");
                return;
            }

            if (lon < -180.0 || lon > 180.0) {
                res.status = 400;
                res.set_content("{\"error\": \"Longitude must be between -180 and 180 degrees\"}", "application/json");
                return;
            }

            if (alt < -1000.0 || alt > 100000.0) {
                res.status = 400;
                res.set_content("{\"error\": \"Altitude must be between -1000 and 100000 feet\"}", "application/json");
                return;
            }

            if (heading < 0.0 || heading > 360.0) {
                res.status = 400;
                res.set_content("{\"error\": \"Heading must be between 0 and 360 degrees\"}", "application/json");
                return;
            }

            if (pitch < -90.0 || pitch > 90.0) {
                res.status = 400;
                res.set_content("{\"error\": \"Pitch must be between -90 and 90 degrees\"}", "application/json");
                return;
            }

            if (cockpit_camera_zoom < 0.0 || cockpit_camera_zoom > 100.0) {
                res.status = 400;
                res.set_content("{\"error\": \"Cockpit camera zoom must be between 0.0 and 100.0\"}", "application/json");
                return;
            }

            // Store the movement data for processing
            {
                std::lock_guard<std::mutex> lock(movement_mutex);
                current_movement_data.latitude = lat;
                current_movement_data.longitude = lon;
                current_movement_data.altitude = alt;
                current_movement_data.heading = heading;
                current_movement_data.pitch = pitch;
                current_movement_data.cockpit_camera_zoom = cockpit_camera_zoom;
                has_movement_data = true;
            }

            // Return success response
            nlohmann::json response = {
                {"status", "success"},
                {"message", "Aircraft movement command received"},
                {"data", {
                    {"lat", lat},
                    {"lon", lon},
                    {"alt", alt},
                    {"heading", heading},
                    {"pitch", pitch},
                    {"cockpit_camera_zoom", cockpit_camera_zoom}
                }}
            };

            res.set_content(response.dump(), "application/json");

        } catch (const std::exception& e) {
            res.status = 400;
            nlohmann::json error_response = {
                {"error", "Invalid JSON format"},
                {"details", e.what()}
            };
            res.set_header("Access-Control-Allow-Origin", "*");
            res.set_header("Access-Control-Allow-Methods", "PUT, GET, POST, OPTIONS");
            res.set_header("Access-Control-Allow-Headers", "*");
            res.set_content(error_response.dump(), "application/json");
        }
    });

    // Health check endpoint
    svr.Get("/health", [](const httplib::Request&, httplib::Response& res) {
        nlohmann::json response = {
            {"status", "healthy"},
            {"service", "aircraft_movement"},
            {"port", 8081}
        };
        res.set_header("Access-Control-Allow-Origin", "*");
        res.set_header("Access-Control-Allow-Methods", "PUT, GET, POST, OPTIONS");
        res.set_header("Access-Control-Allow-Headers", "*");
        res.set_content(response.dump(), "application/json");
    });

    // POST endpoint to reset cockpit view
    svr.Post("/reset", [](const httplib::Request& req, httplib::Response& res) {
        // Set CORS headers first, before any processing
        res.set_header("Access-Control-Allow-Origin", "*");
        res.set_header("Access-Control-Allow-Methods", "PUT, GET, POST, OPTIONS");
        res.set_header("Access-Control-Allow-Headers", "*");
        res.set_header("Access-Control-Max-Age", "86400");
        try {
            nlohmann::json json_data = nlohmann::json::parse(req.body);
            if (!json_data.contains("reset")) {
                res.status = 400;
                res.set_content("{\"error\": \"Missing required field: reset\"}", "application/json");
                return;
            }
            int32_t reset_value = json_data["reset"].get<int32_t>();
            if (reset_value != 1) {
                res.status = 400;
                res.set_content("{\"error\": \"reset must be 1\"}", "application/json");
                return;
            }
            SendCockpitViewReset(reset_value);
            nlohmann::json response = {
                {"status", "success"},
                {"message", "Cockpit view reset command received"},
                {"data", { {"reset", reset_value} }}
            };
            res.set_content(response.dump(), "application/json");
        } catch (const std::exception& e) {
            res.status = 400;
            nlohmann::json error_response = {
                {"error", "Invalid JSON format"},
                {"details", e.what()}
            };
            res.set_header("Access-Control-Allow-Origin", "*");
            res.set_header("Access-Control-Allow-Methods", "PUT, GET, POST, OPTIONS");
            res.set_header("Access-Control-Allow-Headers", "*");
            res.set_content(error_response.dump(), "application/json");
        }
    });

    // CORS preflight handler for reset endpoint
    svr.Options("/reset", [](const httplib::Request&, httplib::Response& res) {
        res.set_header("Access-Control-Allow-Origin", "*");
        res.set_header("Access-Control-Allow-Methods", "PUT, GET, POST, OPTIONS");
        res.set_header("Access-Control-Allow-Headers", "*");
        res.set_header("Access-Control-Max-Age", "86400");
        res.status = 204; // No Content
    });

    std::cout << "Aircraft movement server running on http://localhost:8081" << std::endl;
    std::cout << "  POST/PUT /move - Set aircraft position (JSON: lat, lon, alt, heading, cockpit_camera_zoom)" << std::endl;
    std::cout << "  GET  /health - Health check" << std::endl;

    // Start the server
    if (!svr.listen("0.0.0.0", 8081)) {
        std::cerr << "Failed to start aircraft movement server on port 8081" << std::endl;
    }
}

// Function to start the aircraft movement control system
void StartAircraftMovementControl() {
    // Start the movement processing thread
    std::thread movement_thread(ProcessMovementData);
    
    // Start the HTTP server thread
    std::thread server_thread(AircraftMovementServerThread);

    // Wait for threads to complete (they will run until quit is set)
    if (movement_thread.joinable()) {
        movement_thread.join();
    }
    if (server_thread.joinable()) {
        server_thread.join();
    }
} 

File: src/aircraft_movement.ts
import {
  SimConnectConnection,
  SimConnectConstants,
  RawBuffer,
} from "node-simconnect";
import {
  DEFINITION_3,
  DEFINITION_4,
  MOVEMENT_HTTP_PORT,
  degreesToRadians,
} from "./shared/types.ts";
import { log } from "./logger.ts";

interface AircraftMovementData {
  latitude: number;
  longitude: number;
  altitude: number;
  heading: number;
  pitch: number;
  cockpitCameraZoom: number;
}

interface CockpitViewResetData {
  reset: number;
}

let movementHandle: SimConnectConnection | null = null;

let movementMutex = false;
let currentMovementData: AircraftMovementData = {
  latitude: 0,
  longitude: 0,
  altitude: 0,
  heading: 0,
  pitch: 0,
  cockpitCameraZoom: 1.0,
};
let hasMovementData = false;

let movementThreadQuit = { value: false };

export function setMovementSimConnectHandle(
  handle: SimConnectConnection,
): void {
  movementHandle = handle;
}

export function setupAircraftMovementDefinition(): void {
  if (!movementHandle) {
    log(
      "SimConnect handle is null, cannot set up aircraft movement definition",
    );
    return;
  }

  movementHandle.addToDataDefinition(
    DEFINITION_3,
    "PLANE LATITUDE",
    "degrees",
    4,
  ); // FLOAT64
  movementHandle.addToDataDefinition(
    DEFINITION_3,
    "PLANE LONGITUDE",
    "degrees",
    4,
  );
  movementHandle.addToDataDefinition(DEFINITION_3, "PLANE ALTITUDE", "ft", 4);
  movementHandle.addToDataDefinition(
    DEFINITION_3,
    "PLANE HEADING DEGREES TRUE",
    "radian",
    4,
  );
  movementHandle.addToDataDefinition(
    DEFINITION_3,
    "PLANE PITCH DEGREES",
    "radian",
    4,
  );
  movementHandle.addToDataDefinition(
    DEFINITION_3,
    "COCKPIT CAMERA ZOOM",
    "percentage",
    4,
  );

  movementHandle.addToDataDefinition(
    DEFINITION_4,
    "CAMERA REQUEST ACTION",
    "number",
    1,
  ); // INT32

  log("Aircraft movement data definition (DEFINITION_3) set up successfully");
}

function setAircraftPosition(data: AircraftMovementData): void {
  if (!movementHandle) {
    log("SimConnect handle is null, cannot set aircraft position");
    return;
  }

  const headingRad = degreesToRadians(data.heading);
  const pitchRad = degreesToRadians(data.pitch);

  const buf = new RawBuffer(48);
  buf.writeFloat64(data.latitude);
  buf.writeFloat64(data.longitude);
  buf.writeFloat64(data.altitude);
  buf.writeFloat64(headingRad);
  buf.writeFloat64(pitchRad);
  buf.writeFloat64(data.cockpitCameraZoom);

  movementHandle.setDataOnSimObject(
    DEFINITION_3,
    SimConnectConstants.OBJECT_ID_USER,
    {
      buffer: buf,
      arrayCount: 0,
      tagged: false,
    },
  );
}

function sendCockpitViewReset(resetValue: number): void {
  if (!movementHandle) {
    log("SimConnect handle is null, cannot send cockpit view reset");
    return;
  }

  const buf = new RawBuffer(4);
  buf.writeInt32(resetValue);
  movementHandle.setDataOnSimObject(
    DEFINITION_4,
    SimConnectConstants.OBJECT_ID_USER,
    {
      buffer: buf,
      arrayCount: 0,
      tagged: false,
    },
  );
  log(`Cockpit view reset sent: reset=${resetValue}`);
}

function processMovementData(): void {
  if (movementThreadQuit.value) return;

  let dataToProcess: AircraftMovementData | null = null;

  if (!movementMutex) {
    movementMutex = true;
    if (hasMovementData) {
      dataToProcess = { ...currentMovementData };
      hasMovementData = false;
    }
    movementMutex = false;
  }

  if (dataToProcess) {
    setAircraftPosition(dataToProcess);
  }

  setTimeout(processMovementData, 100);
}

function aircraftMovementServer(): void {
  Bun.serve({
    port: MOVEMENT_HTTP_PORT,
    async fetch(req) {
      const url = new URL(req.url);
      const headers = {
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Methods": "PUT, GET, POST, OPTIONS",
        "Access-Control-Allow-Headers": "*",
        "Access-Control-Max-Age": "86400",
      };

      if (req.method === "OPTIONS") {
        return new Response(null, { status: 204, headers });
      }

      try {
        if (
          (req.method === "POST" || req.method === "PUT") &&
          url.pathname === "/move"
        ) {
          const body = (await req.json()) as Record<string, unknown>;

          const lat = Number(body["lat"]);
          const lon = Number(body["lon"]);
          const alt = Number(body["alt"]);
          const heading = Number(body["heading"]);
          const pitch = Number(body["pitch"]);
          const cockpitCameraZoom = Number(body["cockpit_camera_zoom"]);

          if (
            isNaN(lat) ||
            isNaN(lon) ||
            isNaN(alt) ||
            isNaN(heading) ||
            isNaN(pitch) ||
            isNaN(cockpitCameraZoom)
          ) {
            return new Response(
              JSON.stringify({
                error:
                  "Missing required fields: lat, lon, alt, heading, pitch, cockpit_camera_zoom",
              }),
              {
                status: 400,
                headers: { ...headers, "Content-Type": "application/json" },
              },
            );
          }

          if (lat < -90 || lat > 90) {
            return new Response(
              JSON.stringify({
                error: "Latitude must be between -90 and 90 degrees",
              }),
              {
                status: 400,
                headers: { ...headers, "Content-Type": "application/json" },
              },
            );
          }
          if (lon < -180 || lon > 180) {
            return new Response(
              JSON.stringify({
                error: "Longitude must be between -180 and 180 degrees",
              }),
              {
                status: 400,
                headers: { ...headers, "Content-Type": "application/json" },
              },
            );
          }
          if (alt < -1000 || alt > 100000) {
            return new Response(
              JSON.stringify({
                error: "Altitude must be between -1000 and 100000 feet",
              }),
              {
                status: 400,
                headers: { ...headers, "Content-Type": "application/json" },
              },
            );
          }
          if (heading < 0 || heading > 360) {
            return new Response(
              JSON.stringify({
                error: "Heading must be between 0 and 360 degrees",
              }),
              {
                status: 400,
                headers: { ...headers, "Content-Type": "application/json" },
              },
            );
          }
          if (pitch < -90 || pitch > 90) {
            return new Response(
              JSON.stringify({
                error: "Pitch must be between -90 and 90 degrees",
              }),
              {
                status: 400,
                headers: { ...headers, "Content-Type": "application/json" },
              },
            );
          }
          if (cockpitCameraZoom < 0 || cockpitCameraZoom > 100) {
            return new Response(
              JSON.stringify({
                error: "Cockpit camera zoom must be between 0.0 and 100.0",
              }),
              {
                status: 400,
                headers: { ...headers, "Content-Type": "application/json" },
              },
            );
          }

          movementMutex = true;
          currentMovementData = {
            latitude: lat,
            longitude: lon,
            altitude: alt,
            heading,
            pitch,
            cockpitCameraZoom,
          };
          hasMovementData = true;
          movementMutex = false;

          const response = {
            status: "success",
            message: "Aircraft movement command received",
            data: {
              lat,
              lon,
              alt,
              heading,
              pitch,
              cockpit_camera_zoom: cockpitCameraZoom,
            },
          };
          return new Response(JSON.stringify(response), {
            status: 200,
            headers: { ...headers, "Content-Type": "application/json" },
          });
        }

        if (req.method === "GET" && url.pathname === "/health") {
          const response = {
            status: "healthy",
            service: "aircraft_movement",
            port: MOVEMENT_HTTP_PORT,
          };
          return new Response(JSON.stringify(response), {
            status: 200,
            headers: { ...headers, "Content-Type": "application/json" },
          });
        }

        if (req.method === "POST" && url.pathname === "/reset") {
          const body = (await req.json()) as Record<string, unknown>;
          const resetValue = Number(body["reset"]);
          if (isNaN(resetValue) || resetValue !== 1) {
            return new Response(JSON.stringify({ error: "reset must be 1" }), {
              status: 400,
              headers: { ...headers, "Content-Type": "application/json" },
            });
          }
          sendCockpitViewReset(resetValue);
          const response = {
            status: "success",
            message: "Cockpit view reset command received",
            data: { reset: resetValue },
          };
          return new Response(JSON.stringify(response), {
            status: 200,
            headers: { ...headers, "Content-Type": "application/json" },
          });
        }

        return new Response("Not Found", { status: 404 });
      } catch (e) {
        const errorResponse = {
          error: "Invalid JSON format",
          details: e instanceof Error ? e.message : String(e),
        };
        return new Response(JSON.stringify(errorResponse), {
          status: 400,
          headers: { ...headers, "Content-Type": "application/json" },
        });
      }
    },
  });

  log(
    `Aircraft movement server running on http://localhost:${MOVEMENT_HTTP_PORT}`,
  );
  log(`  POST /move - Set aircraft position`);
  log(`  GET /health - Health check`);
}

export function startAircraftMovementControl(quitFlag: {
  value: boolean;
}): void {
  movementThreadQuit = quitFlag;

  processMovementData();

  aircraftMovementServer();
}
